src/coreclr/debug/ee/riscv64/walker.cpp

```diff
@@ -14,6 +14,152 @@
 
 #ifdef TARGET_RISCV64
 
-// #error "TODO-RISCV64: missing implementation"
+inline uint64_t SignExtend(uint64_t value, unsigned int signbit)
+{
+    _ASSERTE(signbit < 64);
+
+    if (signbit == 63)
+      return value;
+
+    uint64_t sign = value & (1ull << signbit);
+
+    if (sign)
+        return value | (~0ull << signbit);
+    else
+        return value;
+}
+
+inline uint64_t BitExtract(uint64_t value, unsigned int highbit, unsigned int lowbit, bool signExtend = false)
+{
+    _ASSERTE((highbit < 64) && (lowbit < 64) && (highbit >= lowbit));
+    uint64_t extractedValue = (value >> lowbit) & ((1ull << ((highbit - lowbit) + 1)) - 1);
+
+    return signExtend ? SignExtend(extractedValue, highbit - lowbit) : extractedValue;
+}
+
+void NativeWalker::Decode()
+{
+    // Reset so that we do not provide bogus info
+    m_nextIP = NULL;
+    m_skipIP = NULL;
+    m_type = WALK_UNKNOWN;
+
+    if (m_registers == NULL)
+    {
+       // Walker does not use WALK_NEXT
+       // Without registers decoding will work only for handful of instructions
+       return;
+    }
+
+    PT_CONTEXT context = m_registers->pCurrentContext;
+
+    // Fetch first word of the current instruction. If the current instruction is a break instruction, we'll
+    // need to check the patch table to get the correct instruction.
+    PRD_TYPE opcode = CORDbgGetInstruction(m_ip);
+    PRD_TYPE unpatchedOpcode;
+    if (DebuggerController::CheckGetPatchedOpcode(m_ip, &unpatchedOpcode))
+    {
+        opcode = unpatchedOpcode;
+    }
+
+    LOG((LF_CORDB, LL_INFO100000, "RiscV64Walker::Decode instruction at %p, opcode: %x\n", m_ip, opcode));
+
+    // TODO after "C" Standard Extension support implemented, add C.J, C.JAL, C.JR, C.JALR, C.BEQZ, C.BNEZ
+
+    if ((opcode & 0x7f) == 0x6f) // JAL
+    {
+        // J-immediate encodes a signed offset in multiples of 2 bytes
+        //      20       | 19                                               1 | 0
+        // inst[31]/sign | inst[19:12] | inst[20] | inst[30:25] | inst[24:21] | 0
+        uint64_t imm = SignExtend((BitExtract(opcode, 30, 21) << 1) | (BitExtract(opcode, 20, 20) << 11) |
+                                  (BitExtract(opcode, 19, 12) << 12) | (BitExtract(opcode, 31, 31) << 20), 20);
+        uint64_t Rd = BitExtract(opcode, 11, 7);
+
+        m_nextIP = m_ip + imm;
+        // The standard software calling convention uses X1 as the return address register and X5 as an alternate link register.
+        if (Rd == 1 || Rd == 5)
+        {
+            m_skipIP = m_ip + 4;
+            m_type = WALK_CALL;
+        }
+        else
+        {
+            m_skipIP = m_nextIP;
```

Comment:
m_skipIP is set to NULL when walk is WALK_BRANCH for BL & Conditional Branches in ARM64. As I know, skip is "the function you’re currently debugging will be skipped.". I think NULL is right for m_skipIP in this case. Please check. Thank you.

Reply:
I have no idea why, but we have different logic here for arm64 (that have walker code aimed to care about 2 steppers implementation - HW and SW single step) and arm32 (that have only SW single step).
I use same logic as arm32, since it more close to riscv (SW single step only):
https://github.com/dotnet/runtime/blob/45978bd7191fcbb793dc12b162540022063551b6/src/coreclr/debug/ee/arm/walker.cpp#L94-L96
https://github.com/dotnet/runtime/blob/45978bd7191fcbb793dc12b162540022063551b6/src/coreclr/debug/ee/arm/walker.cpp#L115-L117
https://github.com/dotnet/runtime/blob/45978bd7191fcbb793dc12b162540022063551b6/src/coreclr/debug/ee/arm/walker.cpp#L132-L134
https://github.com/dotnet/runtime/blob/45978bd7191fcbb793dc12b162540022063551b6/src/coreclr/debug/ee/arm/walker.cpp#L176-L178
https://github.com/dotnet/runtime/blob/45978bd7191fcbb793dc12b162540022063551b6/src/coreclr/debug/ee/arm/walker.cpp#L277-L279

My point is - stay with same logic as arm32, any way caller look at walk type first and only after that care about variables it really need. Probably this is the point why arm32 and arm64 have different logic here with `m_skipIP`.

Reply:
@jkotas As I know m_skipIP is used for handling skip command in debugger. In debugger, it skips steps for selected functions and files. IMO, `m_skipIP` is not needed for WALK_BRANCH in arm32 as well as riscv64. I investigated `src/coreclr/debug/` and it looks like `m_skipIP` is not used for WALK_BRANCH. Could you please look into it? Thank you.

Reply:
@dotnet/dotnet-diag Could you please help to answer this question?

Reply:
> I investigated src/coreclr/debug/ and it looks like m_skipIP is not used for WALK_BRANCH

The `WALK_BRANCH` case in `TrapStep` falls through to `WALK_CALL`:
https://github.com/dotnet/runtime/blob/35db500cedca4e7061fe1e91390ad1df62bc0f3c/src/coreclr/debug/ee/controller.cpp#L5901-L5911

Which then will read m_skipIP here:
https://github.com/dotnet/runtime/blob/35db500cedca4e7061fe1e91390ad1df62bc0f3c/src/coreclr/debug/ee/controller.cpp#L5978-L5985

Hope this helps!